### PR TITLE
Regression tests: Lifetime

### PR DESCRIPTION
**Pull request by iNinja**
# Regression tests: Lifetime
- Added scenarios for lifetime related validations.

Note: Upcoming PRs will continue to extract the different aspects of regression tests into separate files.

Part of #2711

### CODE REVIEW

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.IdentityModel.TestUtils;
+using Microsoft.IdentityModel.Tokens;
+using Xunit;
+
+namespace Microsoft.IdentityModel.JsonWebTokens.Tests
+{
+    public partial class JsonWebTokenHandlerValidateTokenAsyncTests
```
Comment (jennyf19) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs:13]: why partial? also I can't get these tests to show up in the VS test runner.

Comment (jennyf19) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs:13]: got it work after re-starting

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+                Audience = Default.Audience,
+                Issuer = Default.Issuer,
+                IssuedAt = theoryData.IssuedAt,
+                NotBefore = theoryData.NotBefore,
+                Expires = theoryData.Expires,
+            };
+
+            string jwtString = jsonWebTokenHandler.CreateToken(securityTokenDescriptor);
+
+            TokenValidationResult tokenValidationParametersResult =
```
Comment (jennyf19) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs]: nit: improve the variable name to indicate `legacy`

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+                context.AddDiff($"tokenValidationParametersResult.IsValid != theoryData.ExpectedIsValid");
+
+            if (validationParametersResult.IsSuccess != theoryData.ExpectedIsValid)
+                context.AddDiff($"validationParametersResult.IsSuccess != theoryData.ExpectedIsValid");
+
+            if (theoryData.ExpectedIsValid &&
+                tokenValidationParametersResult.IsValid &&
+                validationParametersResult.IsSuccess)
+            {
+                IdentityComparer.AreEqual(
```
Comment (jennyf19) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs]: do we want to ensure that each specific claim and its value are correct, not just that the claims exist?

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+                    validationParametersResult.UnwrapResult().ClaimsIdentity,
+                    context);
+                IdentityComparer.AreEqual(
+                    tokenValidationParametersResult.Claims,
+                    validationParametersResult.UnwrapResult().Claims,
+                    context);
+            }
+            else
+            {
+                theoryData.ExpectedException.ProcessException(tokenValidationParametersResult.Exception, context);
```
Comment (jennyf19) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs]: Do we need to do more here?  random thoughts: check the type, message, and other properties of the exceptions and ensure that the specific error messages generated are as expected, not just the presence of an exception? or do we think this is enough?

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs [hunk 5]
```diff
@@ -0,0 +1,219 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#nullable enable
```
Comment (FuPingFranco) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs:4]: nit: add a new line after directive to be consistent with Audience regression test PR or vice versa.

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs (hunk 2 above)
Comment (FuPingFranco) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs]: +1

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs (hunk 1 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs:13]: Explained on the other PR, but the main idea is that we split all the regression tests across files for readability and ease to follow. We are separating them by aspect of the validation: lifetime, audience, issuer, replay, actor, signature, token type, encryption, and by using the partial class we can have all the test methods show under the single class in the text explorer.

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs (hunk 3 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs]: We are comparing one set of claims against the other, not just that the claims exist.
If they differ, it gets marked in the context and the test fails.

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs (hunk 4 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs]: At this point we check that:
- The type of the exception is the one we expect
- (if set) the inner exception matches the type we expect
- The `IDX` identifier matches the expectation

What we do not check here:
- The exact contents after the `IDX123123:`
- Custom exception properties

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs (hunk 2 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs]: Renamed to `legacyTokenValidationParametersResult`. Please let me know if you think of something better.

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs (hunk 4 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs]: Added checks for the 2 points mentioned.

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+using System.Threading.Tasks;
+using Microsoft.IdentityModel.TestUtils;
+using Microsoft.IdentityModel.Tokens;
+using Xunit;
+
+namespace Microsoft.IdentityModel.JsonWebTokens.Tests
+{
+    public partial class JsonWebTokenHandlerValidateTokenAsyncTests
+    {
+        [Theory, MemberData(nameof(ValidateTokenAsync_LifetimeTestCases))]
```
Comment (kllysng) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs]:
```suggestion
        [Theory, MemberData(nameof(ValidateTokenAsync_LifetimeTestCases), DisableDiscoveryEnumeration = true)]
```

We want to add this to all of our Theory tests. This enables us to load and run tests faster.

**Review by jennyf19 — APPROVED** [review]
Same for the audiences PR, if we can track all changes to `IDX???` error codes in one GitHub issue, that would be great.

**Review by FuPingFranco — APPROVED** [review]
LGTM

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs (hunk 6 above)
Comment (kllysng) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ValidateTokenAsyncTests.Lifetime.cs]: @iNinja why was this comment resolved?

**Pull request merged**